libbeat/common/kubernetes/metadata/namespace.go

```diff
@@ -0,0 +1,96 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package metadata
+
+import (
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/kubernetes"
+)
+
+const resource = "namespace"
+
+type namespace struct {
+	store    cache.Store
+	resource *Resource
+}
+
+// NewNamespaceMetadataGenerator creates a metagen for namespace resources
+func NewNamespaceMetadataGenerator(cfg *common.Config, namespaces cache.Store) MetaGen {
+	return &namespace{
+		resource: NewResourceMetadataGenerator(cfg),
+		store:    namespaces,
+	}
+}
+
+// Generate generates namespace metadata from a resource object
+func (n *namespace) Generate(obj kubernetes.Resource, opts ...FieldOptions) common.MapStr {
+	_, ok := obj.(*kubernetes.Namespace)
+	if !ok {
+		return nil
+	}
+
+	meta := n.resource.Generate(resource, obj, opts...)
+	// TODO: remove this call when moving to 8.0
```

Comment:
this was added as the proposed event format is:

```
"namespace": {
  "name": <>
  "labels": {
    ...
  }
}
```

this would be a breaking change in 7.x which would cause lot of disruption. The `flatten` method helps retain current structure by using `namespace`, `namespace_labels` and `namespace_uid`.

this hack would need to be removed in 8.0 and use the originally proposed design.